dask/array/core.py

```diff
@@ -2497,8 +2497,8 @@ def broadcast_shapes(*shapes):
         return shapes[0]
     out = []
     for sizes in zip_longest(*map(reversed, shapes), fillvalue=-1):
-        dim = max(sizes)
-        if any(i != -1 and i != 1 and i != dim and not np.isnan(i) for i in sizes):
+        dim = 0 if 0 in sizes else max(sizes)
+        if any(i != -1 and i != 0 and i != 1 and i != dim and not np.isnan(i) for i in sizes):
```

Comment:
Perhaps straighten the left part of this into `i not in (-1, 0, 1, dim)`?

Reply:
Yeah was thinking about that as well.